File: r/day19_part1_2017.R

grid <- readLines("input.txt")
x <- which(strsplit(grid[1], "")[[1]] == "|")[1]
y <- 1

dx <- 0
dy <- 1

letters <- c()

while (TRUE) {
  if (x < 1 || x > nchar(grid[1]) || y < 1 || y > length(grid)) {
    break
  }
  
  cell <- substr(grid[y], x, x)
  
  if (cell == " ") {
    break
  }
  
  if (cell >= "A" && cell <= "Z") {
    letters <- c(letters, cell)
  }
  
  if (cell == "+") {
    if (dx == 0) {
      if (x > 1 && (substr(grid[y], x-1, x-1) == "-" || (substr(grid[y], x-1, x-1) >= "A" && substr(grid[y], x-1, x-1) <= "Z"))) {
        dx <- -1
        dy <- 0
      } else {
        dx <- 1
        dy <- 0
      }
    } else {
      if (y > 1 && (substr(grid[y-1], x, x) == "|" || (substr(grid[y-1], x, x) >= "A" && substr(grid[y-1], x, x) <= "Z"))) {
        dx <- 0
        dy <- -1
      } else {
        dx <- 0
        dy <- 1
      }
    }
  }
  
  x <- x + dx
  y <- y + dy
}

cat(letters, sep = "")

File: ruby/day19_part1_2017.rb
File.open('input.txt', 'r') do |file|
  grid = file.readlines.map(&:chomp).map(&:chars)

  x, y = 0, 0
  grid[0].each_with_index { |cell, i| x = i if cell == '|' }

  dx, dy = 0, 1
  letters = []

  loop do
    break if x < 0 || x >= grid[0].length || y < 0 || y >= grid.length
    cell = grid[y][x]
    break if cell == ' '
    letters << cell if cell.match?(/[A-Z]/)

    if cell == '+'
      if dx.zero?
        dx, dy = x > 0 && (grid[y][x-1] == '-' || grid[y][x-1].match?(/[A-Z]/)) ? -1 : 1, 0
      else
        dx, dy = 0, y > 0 && (grid[y-1][x] == '|' || grid[y-1][x].match?(/[A-Z]/)) ? -1 : 1
      end
    end

    x += dx
    y += dy
  end

  puts letters.join
end